geopandas/tools/overlay.py

```diff
@@ -176,3 +177,158 @@ def overlay(df1, df2, how, use_sindex=True):
+    df1['bbox'] = df1.geometry.apply(lambda x: x.bounds)
+    df1['sidx'] = df1.bbox.apply(lambda x:list(spatial_index.intersection(x)))
+    # Cretae differences
+    df1['new_g'] = df1.apply(lambda x: reduce(lambda x, y: x.difference(y).buffer(0), 
+                             [x.geometry]+list(df2.iloc[x.sidx].geometry)) , axis=1)
+    df1.geometry = df1.new_g
+    df1 = df1.loc[df1.geometry.is_empty==False].copy()
+    df1.drop(['bbox', 'sidx', 'new_g'], axis=1, inplace=True)
+    df1 = df1.reset_index(drop=True)
+    return df1
+
+def overlay_symmetric_diff(df1, df2):
+    ''''
+    Overlay Symmetric Difference operation used in overlay function
+    '''
+    df1['idx1'] = df1.index
+    df2['idx2'] = df2.index
+    df1['idx2'] = np.nan
+    df2['idx1'] = np.nan
+    dfsym = df1.merge(df2, on=['idx1','idx2'], how='outer', suffixes=['_1','_2'])
+    dfsym['geometry'] = dfsym.geometry_1
+    dfsym.loc[dfsym.geometry_2.isnull()==False, 'geometry'] = dfsym.loc[dfsym.geometry_2.isnull()==False, 'geometry_2']
+    dfsym.drop(['geometry_1', 'geometry_2'], axis=1, inplace=True)
+    dfsym = GeoDataFrame(dfsym, columns=dfsym.columns, crs=df1.crs)
+    # Spatial Index to create intersections
+    spatial_index = dfsym.sindex
+    dfsym['bbox'] = dfsym.geometry.apply(lambda x: x.bounds)
+    dfsym['sidx'] = dfsym.bbox.apply(lambda x:list(spatial_index.intersection(x)))
+    dfsym['idx'] = dfsym.index.values
+    dfsym.apply(lambda x: x.sidx.remove(x.idx), axis=1)
+    dfsym['new_g'] = dfsym.apply(lambda x: reduce(lambda x, y: x.difference(y).buffer(0), 
+                     [x.geometry]+list(dfsym.iloc[x.sidx].geometry)) , axis=1)
+    dfsym.geometry = dfsym.new_g
+    dfsym = dfsym.loc[dfsym.geometry.is_empty==False].copy()
+    dfsym.drop(['bbox', 'sidx', 'idx', 'idx1','idx2', 'new_g'], axis=1, inplace=True)
+    dfsym = dfsym.reset_index(drop=True)
+    return dfsym
+
+def overlay(df1, df2, how='intersection', make_valid=True, reproject=True, use_sindex=None, **kwargs):
+    """Perform spatial overlay between two polygons.
+
+    Currently only supports data GeoDataFrames with polygons.
+    Implements several methods that are all effectively subsets of
+    the union.
+
+    Parameters
+    ----------
+    df1 : GeoDataFrame with MultiPolygon or Polygon geometry column
+    df2 : GeoDataFrame with MultiPolygon or Polygon geometry column
+    how : string
+        Method of spatial overlay: 'intersection', 'union',
+        'identity', 'symmetric_difference' or 'difference'.
+    reproject : boolean, default True
+        If GeoDataFrames do not have same projection, reproject
+        df2 to same projection of df1 before performing overlay
+
+    Returns
+    -------
+    df : GeoDataFrame
+        GeoDataFrame with new set of polygons and attributes
+        resulting from the overlay
+
+    """
+    if use_sindex is not None:
+        print('use_sindex is deprecated. If you are trying to use the old "overlay" function use "overlay_slow".')
```

Comment:
Can you make this also into an actual Warning? (you can use `FutureWarning`).

Just to check: you removed this option because your current implementation would not be possible without an `sindex`? (or only much slower? which might still be a reason to require it, to avoid people being surprised about the slowness)